File: frontend/src/types/streaming.ts
export type AgentStatus = "running" | "done" | "error";

export interface AgentLogEvent {
  event: "agent_log";
  agent: string;
  label: string;
  message: string;
  status: AgentStatus;
  data?: Record<string, unknown>;
}

export interface HeartbeatEvent {
  event: "heartbeat";
}

export interface ErrorEvent {
  event: "error";
  agent?: string;
  message: string;
}

export type SSEEvent = AgentLogEvent | HeartbeatEvent | ErrorEvent;

File: backend/src/models/schemas/streaming.py
from typing import Any, Literal

from pydantic import BaseModel

AgentStatus = Literal["running", "done", "error"]


class AgentLogEvent(BaseModel):
    """Single SSE event payload for one agent log line."""

    event: Literal["agent_log"] = "agent_log"
    agent: str
    label: str
    message: str
    status: AgentStatus = "running"
    data: Any = None


class ReportFinalEvent(BaseModel):
    """Final SSE event carrying the complete LeaseLensReport."""

    event: Literal["report_final"] = "report_final"
    report: dict


class HeartbeatEvent(BaseModel):
    """Keep-alive SSE event."""

    event: Literal["heartbeat"] = "heartbeat"


class ErrorEvent(BaseModel):
    """SSE error event."""

    event: Literal["error"] = "error"
    agent: str | None = None
    message: str
